server/hwapi/endpoints/certification/logic.py

```diff
@@ -46,27 +51,203 @@ def is_certified(system_info: CertificationStatusRequest, db: Session):
+            model=device.name, vendor=device.vendor.name, identifier=device.identifier
+        ),
+    )
+
+    pci_devices = _get_matching_devices(
+        db,
+        devices_query,
+        system_info.pci_peripherals or [],
+        repository.find_matching_pci_device,
+        data_validators.PCIPeripheralValidator,
+        lambda device, _: data_validators.PCIPeripheralValidator(
+            name=device.name, vendor=device.vendor.name, pci_id=device.identifier
+        ),
+    )
+
+    usb_devices = _get_matching_devices(
+        db,
+        devices_query,
+        system_info.usb_peripherals or [],
+        repository.find_matching_usb_device,
+        data_validators.USBPeripheralValidator,
+        lambda device, _: data_validators.USBPeripheralValidator(
+            name=device.name, vendor=device.vendor.name, usb_id=device.identifier
+        ),
+    )
+
+    # If there are no matching devices of any type, return False
+    if not any(
+        [
+            gpus,
+            cpus,
+            network_devices,
+            wireless_devices,
+            audio_devices,
+            capture_devices,
+            pci_devices,
+            usb_devices,
+        ]
+    ):
+        return False, None
+
+    return True, RelatedCertifiedSystemExistsResponse(
+        architecture=system_info.architecture,
+        board=(
+            data_validators.BoardValidator(
+                manufacturer=board.vendor.name,
+                version=board.version,
+                product_name=board.name,
+            )
+            if board
+            else None
+        ),
+        processor=cpus,
+        gpu=gpus,
+        audio=audio_devices,
+        video=capture_devices,
+        network=network_devices,
+        wireless=wireless_devices,
+        pci_peripherals=pci_devices,
+        usb_peripherals=usb_devices,
+    )
+
+
+def _get_matching_devices(
+    db: Session,
+    devices_query: Query,
+    device_validator_list: list,
+    find_matching_function: Callable[[Session, Query, Any], Device | None],
+    device_validator_class: type[BaseModel],
+    validator_mapping: Callable[[Device, Any], Any],
+):
+    """
+    General-purpose function to get matching devices.
+
+    :param db: SQLAlchemy database session.
+    :param devices_query: Devices query object.
+    :param device_validator_list: List of device validator objects.
+    :param find_matching_function: The function that retrieves matching devices.
+    :param device_validator_class: The validator class to return as response.
+    :param validator_mapping: Mapping function to transform devices to validator objects.
+    :return: A list of matching device validator objects.
+    """
+    devices = []
+    for device_info in device_validator_list:
+        device = find_matching_function(db, devices_query, device_info)
+        if device is None:
+            continue
+        devices.append(validator_mapping(device, device_info))
+    return devices
```

Comment:
This function is too complex. Personally, I don't know exactly what it does. Perhaps it's doing too many things (as I can tell by the number of parameters it takes). I'd would've tried to find a way to simplify it, but I didn't really understand it